### ISSUE TITLE
Low confidence

### ISSUE THREAD

**Issue by betterhalfwzm** [zylo117/Yet-Another-EfficientDet-Pytorch#157]
Has anyone trained custom dataset to get high confidence?  Need help!!!
Thx!!!!!!!!!!

**Comment by yytang2012** [comment]
The confidence is less than 0.1 in most cases on my custom datasets

**Comment by betterhalfwzm** [comment]
> The confidence is less than 0.1 in most cases on my custom datasets

Mine is almost the same. It's not normal. Has someone solved the problem?

**Comment by Cli98** [comment]
So does mine

> > The confidence is less than 0.1 in most cases on my custom datasets
> 
> Mine is almost the same. It's not normal. Has someone solved the problem?

**Comment by zylo117** [comment]
Did you test on training images? Is their confidence high?

**Comment by betterhalfwzm** [comment]
> Did you test on training images? Is their confidence high?

I test, the same, low confidence.

**Comment by zylo117** [comment]
Can you train on shape dataset? The one in the tutorial?
If so, it's likely your dataset is too difficult and it's still underfitting.

**Comment by betterhalfwzm** [comment]
> Can you train on shape dataset? The one in the tutorial? It takes only 30 minutes to get 70+ mAP.
> If so, it's likely your dataset is too difficult and it's still underfitting.
> Is your category id starts at 1, not 0?

Yes, category id is 1, not 0

**Comment by bzzonly** [comment]
my custom datasets also get results 0. And i use train_set to val , the result is normal. My category is 1. What reaso do you think? 

> Did you test on training images? Is their confidence high?

**Comment by zylo117** [comment]
Actually, I was wondering, what do you mean by confidence? Do you mean mAP or just average confidence of the boxes?

And loss is just loss, it's not mAP. Lower loss doesn't mean that it's good. You still need to run coco_eval to get a objective result.

The coco_eval will generate a temporary file call val_xxxxx.json, you should manually remove it or the next time you run coco_eval, the result will be the same. This issue is fixed if you are using the latest code.

**Comment by betterhalfwzm** [comment]
> Actually, I was wondering, what do you mean by confidence? Do you mean mAP or just average confidence of the boxes?
> 
> And loss is just loss, it's not mAP. Lower loss doesn't mean that it's good. You still need to run coco_eval to get a objective result.
> 
> The coco_eval will generate a temporary file called val_xxxxx.json, you should manually remove it or the next time you run coco_eval, the result will be the same. This issue is fixed if you are using the latest code.

 Just average confidence of the boxes.  official Tensorflow evey image has a high confidence 0.7-0.9, but your code only has 0.001-0.2.

**Comment by zylo117** [comment]
wait a minute, do you mean you try to infer images with my coco pretrained weights and have low confidence?

**Comment by betterhalfwzm** [comment]
> wait a minute, do you mean you try to infer images with my coco pretrained weights and have low confidence?

No, my trained model on my dataset.

**Comment by zylo117** [comment]
And you have trained another model on tf's effdet?

**Comment by betterhalfwzm** [comment]
> And you have trained another model on tf's effdet?

Yes, i have trained tf model,and confidence is high.  high is tf  ,low is your code 
![image](https://user-images.githubusercontent.com/20449990/79847229-fe98b880-83f1-11ea-8620-f85771932e74.png)
![image](https://user-images.githubusercontent.com/20449990/79847241-05273000-83f2-11ea-90d3-f63fc304f254.png)

**Comment by zylo117** [comment]
I see, did you force input size to be original image size or not? ASAIK, the demo of tf's does not downsample image in inference.

**Comment by zylo117** [comment]
For example. 
The first one is 1920, for demo propose, almost the same as tf's.
![img_inferred_d0_this_repo](https://user-images.githubusercontent.com/5842006/79849010-7831a600-83f4-11ea-96dd-95342ac6a219.jpg)

The second one is 512, standard D0 input,
![img_inferred_d0_this_repo_0](https://user-images.githubusercontent.com/5842006/79849120-a3b49080-83f4-11ea-9903-9f42110ab032.jpg)

Actually the first one is kind of cheating, but tf did it, so I have to do it. You need to make sure it is test under the same resolution.

**Comment by betterhalfwzm** [comment]
> For example.
> The first one is 1920, for demo propose, almost the same as tf's.
> ![img_inferred_d0_this_repo](https://user-images.githubusercontent.com/5842006/79849010-7831a600-83f4-11ea-96dd-95342ac6a219.jpg)
> 
> The second one is 512, standard D0 input,
> ![img_inferred_d0_this_repo_0](https://user-images.githubusercontent.com/5842006/79849120-a3b49080-83f4-11ea-9903-9f42110ab032.jpg)
> 
> Actually the first one is kind of cheating, but tf did it, so I have to do it. You need to make sure it is tested under the same resolution.

val input size both set 512

**Comment by zylo117** [comment]
well then it should be training issue. It is still not converged

**Comment by purity77** [comment]
I also have the same problem with not converged

**Comment by gtgtgt1117** [comment]
> > Actually, I was wondering, what do you mean by confidence? Do you mean mAP or just average confidence of the boxes?
> > And loss is just loss, it's not mAP. Lower loss doesn't mean that it's good. You still need to run coco_eval to get a objective result.
> > The coco_eval will generate a temporary file called val_xxxxx.json, you should manually remove it or the next time you run coco_eval, the result will be the same. This issue is fixed if you are using the latest code.
> 
> Just average confidence of the boxes. official Tensorflow evey image has a high confidence 0.7-0.9, but your code only has 0.001-0.2.

hello @betterhalfwzm could you tell me how to infer the image? i try to infer the image,but there isn't results .And i can't save the model and use the saved_model_infer.

**Comment by betterhalfwzm** [comment]
@gtgtgt1117 efficientdet_test.py threshold = 0.01

**Comment by mad-fogs** [comment]
I have got same question on my custom dataset.
Especially on certain not so "hard" samples, large object got very low confidence and then they will be ignored there.
I dont mean this is a bug, but this really never happened in previous projects.

**Comment by zylo117** [comment]
I think maybe focal_loss need to be optimize? Focal loss function makes converging much slower. So when not converging, the confidence is low.

**Comment by betterhalfwzm** [comment]
> I think maybe focal_loss need to be optimize? Focal loss function makes converging much slower. So when not converging, the confidence is low.

I think there is something wrong. Another code trains my dataset, and it doesn't matter.
https://github.com/thuyngch/ATSS-EfficientDet-PyTorch

**Comment by mad-fogs** [comment]
Hi gus,

I do tried d0~d4 on my datasets, and i found the auther @zylo117  is right.
low confidence seems like resulted from insufficient traing steps.
I trained the d0 d1 d2 with more epoches than d4 and the confidence comes to be normal .

**Comment by SubramanianKrish** [comment]
@mad-fogs Could you share details of your dataset used for training the d0? specifically number of training images, batch size, num epochs. I'm seeing low confidence boxes all over my val set too!
